dashboard/signals.py

```diff
@@ -25,27 +38,61 @@ def precreate_programenrollment(sender, instance, **kwargs):  # pylint: disable=
+            if CachedEnrollment.active_count(user, program) <= 1:  # theoretically this cannot be <1, but just in case
                 ProgramEnrollment.objects.filter(
                     user=user,
                     program=program
                 ).delete()
 
 
-@receiver(post_save, sender=CachedEnrollment, dispatch_uid="update_programenrollment")
-def create_programenrollment(sender, instance, **kwargs):  # pylint: disable=unused-argument
+@receiver(post_save, sender=CachedEnrollment, dispatch_uid="cachedenrollment_post_save")
+def handle_update_enrollment(sender, instance, **kwargs):  # pylint: disable=unused-argument
     """
-    Signal handler to create Program enrollment when the CachedEnrollment table is updated
+    Create ProgramEnrollment when a CachedEnrollment is created/updated, and update the index
     """
     if instance.data is not None:
-        ProgramEnrollment.objects.get_or_create(
+        program_enrollment = ProgramEnrollment.objects.get_or_create(
             user=instance.user,
             program=instance.course_run.course.program
-        )
+        )[0]
+        index_program_enrolled_users.delay([program_enrollment])
+
+
+@receiver(post_save, sender=CachedCertificate, dispatch_uid="cachedcertificate_post_save")
+def handle_update_certificate(sender, instance, **kwargs):  # pylint: disable=unused-argument
+    """When a CachedCertificate model is updated, update index """
+    if instance.data is not None:
+        program_enrollment = ProgramEnrollment.objects.get_or_create(
+            user=instance.user,
+            program=instance.course_run.course.program
+        )[0]
+        index_program_enrolled_users.delay([program_enrollment])
+
+
+@receiver(post_delete, sender=CachedEnrollment, dispatch_uid="cachedenrollment_post_delete")
+def handle_delete_enrollment(sender, instance, **kwargs):  # pylint: disable=unused-argument
+    """Update index when CachedEnrollment model instance is deleted."""
+    user = instance.user
+    program = instance.course_run.course.program
+    program_enrollment = ProgramEnrollment.objects.filter(user=user, program=program).first()
+    if program_enrollment and CachedEnrollment.active_count(user, program) == 0:
+        program_enrollment.delete()
+    if not program_enrollment or not program_enrollment.id:
+        index_users.delay([user])
+    else:
+        index_program_enrolled_users.delay([program_enrollment])
+
+
+@receiver(post_delete, sender=CachedCertificate, dispatch_uid="cachedcertificate_post_delete")
+def handle_delete_certificate(sender, instance, **kwargs):  # pylint: disable=unused-argument
+    """Update index when CachedCertificate model instance is deleted."""
+    user = instance.user
+    program = instance.course_run.course.program
+    program_enrollment = ProgramEnrollment.objects.filter(user=user, program=program).first()
+    if not program_enrollment:
+        index_users.delay([user])
```

Comment:
keep in mind that this is looking for a `ProgramEnrollment` for a User _and_ a `Program`. A User could have other ProgramEnrollments. your comment does point out an issue, though. i had this code in place to handle cases where the ProgramEnrollment had already been deleted by an earlier signal, but that in itself should trigger a reindexing. ill get rid of this and change up the enrollment signal as well